File: Beecrowd_Parte1/A_Noiva_Do_Trevo_3149.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_Q 1000

// Estrutura para armazenar os relatos
struct Relato {
    char nome[101]; // Nome do morador
    int horario;    // Horário do avistamento em minutos desde a meia-noite
};

// Função para converter o horário HH:MM para minutos desde a meia-noite
int parseTime(char *hora) {
    int horas, minutos;
    sscanf(hora, "%d:%d", &horas, &minutos);
    return horas * 60 + minutos;
}

// Função de comparação para ordenar os relatos por horário
int compararRelatos(const void *a, const void *b) {
    struct Relato *relatoA = (struct Relato *)a;
    struct Relato *relatoB = (struct Relato *)b;
    return relatoA->horario - relatoB->horario;
}

int main() {
    int P, Q;
    struct Relato relatos[MAX_Q];

    // Leitura de P e Q
    scanf("%d %d", &P, &Q);

    // Leitura dos relatos
    for (int i = 0; i < Q; i++) {
        char hora[6], nome[101];
        scanf("%s %[^\n]", hora, nome);

        // Converte o horário para minutos desde a meia-noite
        int minutos = parseTime(hora);

        // Verifica se está dentro do intervalo próximo à meia-noite
        if (minutos <= P || minutos >= 1440 - P) {
            strcpy(relatos[i].nome, nome);
            relatos[i].horario = minutos;
        } else {
            // Não armazena relatos fora do intervalo
            i--;
            Q--;
        }
    }

    // Ordena os relatos pelo horário
    qsort(relatos, Q, sizeof(struct Relato), compararRelatos);

    // Saída dos nomes dos moradores ordenados pelo horário do relato
    for (int i = 0; i < Q; i++) {
        printf("%s\n", relatos[i].nome);
    }
    return 0;
}

File: Beecrowd_Parte1/A_Noiva_Do_Trevo_3149.java
import java.util.*;

public class A_Noiva_Do_Trevo_3149 {

    static class Relato implements Comparable<Relato> {
        String nome;
        int horario;

        public Relato(String nome, int horario) {
            this.nome = nome;
            this.horario = horario;
        }

        @Override
        public int compareTo(Relato other) {
            return Integer.compare(this.horario, other.horario);
        }
    }

    // Função para converter o horário HH:MM para minutos desde a meia-noite
    static int parseTime(String hora) {
        int horas = Integer.parseInt(hora.substring(0, 2));
        int minutos = Integer.parseInt(hora.substring(3));
        return horas * 60 + minutos;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int P = sc.nextInt(); // minutos para mais e para menos da meia-noite
        int Q = sc.nextInt(); // quantidade de pessoas que relataram

        List<Relato> relatos = new ArrayList<>();

        // Leitura dos relatos
        for (int i = 0; i < Q; i++) {
            String H = sc.next(); // horário do relato
            String N = sc.nextLine().trim(); // nome do morador

            int minutos = parseTime(H);

            // Verifica se está dentro do intervalo próximo à meia-noite
            if (minutos <= P || minutos >= 1440 - P) {
                relatos.add(new Relato(N, minutos));
            }
        }

        // Ordena os relatos pelo horário
        Collections.sort(relatos);

        // Saída dos nomes dos moradores ordenados pelo horário do relato
        for (Relato relato : relatos) {
            System.out.println(relato.nome);
        }
    }
}
